demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,1087 @@
+            xStatus = prvRequestS3ObjectRange( &xRequestInfo,
+                                               xRequestQueue,
+                                               xCurByte,
+                                               xCurByte + xNumReqBytes - 1 );
+
+            /* Exit loop if the request was not successfully enqueued. */
+            if( xStatus != pdPASS )
+            {
+                break;
+            }
+
+            /* Update the starting byte for the next iteration.*/
+            xCurByte += xNumReqBytes;
+
+            /* If the number of bytes left to download is less than the
+             * pre-defined constant xNumReqBytes, set xNumReqBytes to equal the
+             * accurate number of remaining bytes left to download. */
+            if( ( xFileSize - xCurByte ) < xNumReqBytes )
+            {
+                xNumReqBytes = xFileSize - xCurByte;
+            }
+
+            /* If the request was successfully enqueued, we expect a
+             * corresponding response. */
+            xResponseCount += 1;
+        }
+
+        /* Retrieve response from the response queue, if available. */
+        if( xResponseCount > 0 )
+        {
+            if( xQueueReceive( xResponseQueue, &xResponseItem, httpexampleDEMO_TICKS_TO_WAIT ) != pdFAIL )
+            {
+                /* Ensure that the buffer pointer is accurate after being copied from the queue. */
+                xResponseItem.xResponse.pBuffer = xResponseItem.ucResponseBuffer;
+
+                LogInfo( ( "The main task retrieved a server response from the response queue." ) );
+                LogDebug( ( "Response Headers:\n%.*s",
+                            ( int32_t ) xResponseItem.xResponse.headersLen,
+                            xResponseItem.xResponse.pHeaders ) );
+                LogDebug( ( "Response Status:\n%u",
+                            xResponseItem.xResponse.statusCode ) );
+                LogInfo( ( "Response Body:\n%.*s\n",
+                           ( int32_t ) xResponseItem.xResponse.bodyLen,
+                           xResponseItem.xResponse.pBody ) );
+
+                /* Check for a partial content status code (206), indicating a
+                 * successful server response. */
+                if( xResponseItem.xResponse.statusCode != httpexampleHTTP_STATUS_CODE_PARTIAL_CONTENT )
+                {
+                    LogError( ( "Received response with unexpected status code: %d", xResponseItem.xResponse.statusCode ) );
+                    xStatus = pdFAIL;
+                    break;
+                }
+
+                /* Reset the wait counter every time a response is received. */
+                ulWaitCounter = 0;
+                xResponseCount -= 1;
+            }
+            /* Check for a notification from the HTTP task about an HTTP send failure. */
+            else if( prvCheckNotification( &ulNotification, httpexampleHTTP_SEND_ERROR, true ) != false )
```

Comment:
Can the HTTP task send a dummy response with `xResponseItem.xResponse.statusCode` set to some error code indicating send failed?

Reply:
Yes definitely, it could, but is there an issue with this method of sending notifications between tasks? (I ask because they are used more frequently in the P3 demo, which adds a couple of additional tasks).

Reply:
It is more like a simplification which can be done later.